cpp/plugins/cucim.kit.cumed/src/cumed/cumed.cpp

```diff
@@ -0,0 +1,419 @@
+    {
+        throw std::runtime_error("out_metadata_desc shouldn't be nullptr!");
+    }
+    cucim::io::format::ImageMetadata& out_metadata =
+        *reinterpret_cast<cucim::io::format::ImageMetadata*>(out_metadata_desc->handle);
+
+
+    //
+    // Metadata Setup
+    //
+
+    // Note: int-> uint16_t due to type differences between ImageMetadataDesc.ndim and DLTensor.ndim
+    const uint16_t ndim = 3;
+    auto& resource = out_metadata.get_resource();
+
+    std::string_view dims{ "YXC" };
+
+    std::pmr::vector<int64_t> shape({ 256, 256, 3 }, &resource);
+
+    DLDataType dtype{ kDLUInt, 8, 1 };
+
+    // Assume RGB
+    std::pmr::vector<std::string_view> channel_names(
+        { std::string_view{ "R" }, std::string_view{ "G" }, std::string_view{ "B" } }, &resource);
+
+    std::pmr::vector<float> spacing(&resource);
+    spacing.reserve(ndim);
+    spacing.insert(spacing.end(), ndim, 1.0);
+
+    std::pmr::vector<std::string_view> spacing_units(&resource);
+    spacing_units.reserve(ndim);
+    spacing_units.emplace_back(std::string_view{ "pixel" });
+    spacing_units.emplace_back(std::string_view{ "pixel" });
+    spacing_units.emplace_back(std::string_view{ "color" });
+
+    std::pmr::vector<float> origin({ 0.0, 0.0, 0.0 }, &resource);
+    // Direction cosines (size is always 3x3)
+    // clang-format off
+    std::pmr::vector<float> direction({ 1.0, 0.0, 0.0,
+                                        0.0, 1.0, 0.0,
+                                        0.0, 0.0, 1.0}, &resource);
+    // clang-format on
+
+    // The coordinate frame in which the direction cosines are measured (either 'LPS'(ITK/DICOM) or 'RAS'(NIfTI/3D
+    // Slicer))
+    std::string_view coord_sys{ "LPS" };
+
+    size_t level_count = 1;
+    const uint16_t level_ndim = 2; // {'X', 'Y'}
+    std::pmr::vector<int64_t> level_dimensions(&resource);
+    level_dimensions.reserve(level_count * 2);
+    for (size_t i = 0; i < level_count; ++i)
+    {
+        level_dimensions.emplace_back(256);
+        level_dimensions.emplace_back(256);
+    }
+
+    std::pmr::vector<float> level_downsamples(&resource);
+    float orig_width = static_cast<float>(shape[1]);
+    float orig_height = static_cast<float>(shape[0]);
+    for (size_t i = 0; i < level_count; ++i)
+    {
+        level_downsamples.emplace_back(1.0);
+    }
+
+    std::pmr::vector<uint32_t> level_tile_sizes(&resource);
+    level_tile_sizes.reserve(level_count * 2);
+    for (size_t i = 0; i < level_count; ++i)
+    {
+        level_tile_sizes.emplace_back(256);
+        level_tile_sizes.emplace_back(256);
+    }
+
+    const size_t associated_image_count = 0;
+    std::pmr::vector<std::string_view> associated_image_names(&resource);
+
+    std::string_view raw_data{ "" };
+
+    // Dynamically allocate memory for json_data (need to be freed manually);
+    const std::string& json_str = std::string{};
+    char* json_data_ptr = static_cast<char*>(cucim_malloc(json_str.size() + 1));
```

Comment:
Is there a corresponding `free` somewhere? Would it be possible to use RAII to handle this allocation?

Reply:
Yes, it is at the destructor of CuImage class.

https://github.com/rapidsai/cucim/blob/95e44833752209b548c1b5e7aa40217d0187909b/cpp/src/cuimage.cpp#L236-L241

I have created an issue to handle general resource handling issues and will follow up on the improvement through the issue.
- https://github.com/rapidsai/cucim/issues/135